### PR TITLE
test: add test covergate to PushDividendPayment procedure

### PR DESCRIPTION
**Pull request by shuffledex**

### CODE REVIEW

File: src/procedures/__tests__/PushDividendPayment.ts [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+import { ProcedureType, DividendType, ErrorCode } from '~/types';
+import { PolymathError } from '~/PolymathError';
+import { mockFactories } from '~/testUtils/mockFactories';
+import { Factories } from '../../Context';
+import { SecurityToken, DividendDistribution } from '~/entities';
+
+const params = {
+  symbol: 'TEST1',
+  dividendIndex: 0,
+  shareholderAddresses: ['0x01'],
```
Comment (victorwiebe) [inline src/procedures/__tests__/PushDividendPayment.ts]: Fede, I just have one major comment here. 
Checking the test metrics- there seems to be something else in the logic that you can test (branch coverage 80%) 
Uncovered lines are 98, 118. What are these lines?
Line 98:
```
    if (shareholderAddresses) {
```
And Line 118: 
```
index < shareholderAddressChunks.length - 1
```
This code makes me dizzy too, so I'd suggest consulting Jere if things are not clear.

What I _think_ is missing here is that you need to test the path the procedure takes when the shareholderAddresses are empty. Well, to be more specific, in line 98 we want to make sure that the tests take both paths, 
`shareholderAddresses` and `!shareholderAddresses` . I think that this is related as well to 118. I'm a little lost in respect to what index and shareholderAddressChunks exactly do, but I have a strong feeling that the test metrics are complaining because you are not testing both paths here (I think the case where `undefined` is also missing from your current test suite)
```  
index < shareholderAddressChunks.length - 1
              ? undefined
              : createPushDividendPaymentResolver(
                  factories,
                  symbol,...
```

File: src/procedures/__tests__/PushDividendPayment.ts [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+        addTransactionSpy
+          .getCall(0)
+          .calledWith(
+            ethDividendsMock.getMockInstance().pushDividendPaymentToAddresses
+          )
+      ).toEqual(true);
+      expect(addTransactionSpy.callCount).toEqual(1);
+    });
+
+    test('should successfully refresh ERC20 dividends distribution factory', async () => {
```
Comment (monitz87) [inline src/procedures/__tests__/PushDividendPayment.ts]: The thing being refreshed is not the factory, but the dividends distribution

The description should be something like 'should refresh the ERC20 dividend distribution for which payments were pushed'

Same for the next one but with ETH

File: src/procedures/__tests__/PushDividendPayment.ts [hunk 3]
```diff
@@ -0,0 +1,312 @@
+import { ImportMock, MockManager } from 'ts-mock-imports';
+import { spy, restore } from 'sinon';
+import * as contractWrappersModule from '@polymathnetwork/contract-wrappers';
+import * as contextModule from '../../Context';
+import * as wrappersModule from '../../PolymathBase';
+import * as tokenFactoryModule from '../../testUtils/MockedTokenFactoryModule';
+import * as dividendFactoryModule from '~/entities/factories/DividendDistributionFactory';
+import * as pushDividendPaymentModule from '../../procedures/PushDividendPayment';
+import { PushDividendPayment } from '../../procedures/PushDividendPayment';
+import { Procedure } from '~/procedures/Procedure';
```
Comment (victorwiebe) [inline src/procedures/__tests__/PushDividendPayment.ts]: Prefer ../Procedure instead of ~/

File: src/procedures/__tests__/PushDividendPayment.ts [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+import { ProcedureType, DividendType, ErrorCode } from '~/types';
```
Comment (victorwiebe) [inline src/procedures/__tests__/PushDividendPayment.ts]: Prefer ../../Types instead of ~/

File: src/procedures/__tests__/PushDividendPayment.ts [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+import { PolymathError } from '~/PolymathError';
```
Comment (victorwiebe) [inline src/procedures/__tests__/PushDividendPayment.ts]: Prefer ../../PolymathError instead of ~/

File: src/procedures/__tests__/PushDividendPayment.ts [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+import { mockFactories } from '~/testUtils/mockFactories';
```
Comment (victorwiebe) [inline src/procedures/__tests__/PushDividendPayment.ts]: Prefer ../../testUtils/mockFactories instead of ~/

File: src/procedures/__tests__/PushDividendPayment.ts [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+import { Factories } from '../../Context';
+import { SecurityToken, DividendDistribution } from '~/entities';
```
Comment (victorwiebe) [inline src/procedures/__tests__/PushDividendPayment.ts]: prefer ../../entities instead of ~/

File: src/procedures/__tests__/PushDividendPayment.ts [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+    test('should extend procedure and have PushDividendPayment type', async () => {
+      target = new PushDividendPayment(params, contextMock.getMockInstance());
+      expect(target instanceof Procedure).toBe(true);
+      expect(target.type).toBe(ProcedureType.PushDividendPayment);
+    });
+  });
+
+  describe('PushDividendPayment', () => {
+    test('should throw if there is no valid security token being provided', async () => {
+      // Instantiate PushDividendPayment with incorrect security symbol
```
Comment (victorwiebe) [inline src/procedures/__tests__/PushDividendPayment.ts]: I'm not sure this comment reflects the line below it. You can move the target = ... instantiation in the beforeEach if you want, your choice

File: src/procedures/__tests__/PushDividendPayment.ts [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+      // Real call
+      await expect(target.prepareTransactions()).rejects.toThrowError(
+        new PolymathError({
+          code: ErrorCode.ProcedureValidationError,
+          message: "Dividends of the specified type haven't been enabled",
+        })
+      );
+    });
+
+    test('should add a transaction to push a dividend payment of an attached ERC20 dividends module', async () => {
```
Comment (victorwiebe) [inline src/procedures/__tests__/PushDividendPayment.ts]: Maybe, "should add a transaction to push a dividend payment *for* an attached ERC20 dividends module"

File: src/procedures/__tests__/PushDividendPayment.ts [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+      const addTransactionSpy = spy(target, 'addTransaction');
+
+      // Real call
+      await target.prepareTransactions();
+
+      // Verifications
+      expect(
+        addTransactionSpy
+          .getCall(0)
+          .calledWith(erc20DividendsMock.getMockInstance().pushDividendPaymentToAddresses)
```
Comment (victorwiebe) [inline src/procedures/__tests__/PushDividendPayment.ts]: This is now only partially correct. You are missing 
`erc20DividendsMock.mock('pushDividendPaymentToAddresses', Promise.resolve('PushDividendPaymentToAddresses'));` -> added somewhere before target.prepareTransactions

Where the resolve has a unique string. This is as per what I mentioned in slack today, let me know if you have more questions.

To test this, as the file is right now, you could change erc20DividendsMock on this line to ethDividendsMock, and it would still pass. That is because without the mock. the functions are all `undefined` type. If you add the code I mention above, you will see that you are no longer able to pass anything else into the calledWith, *only* this.

File: src/procedures/__tests__/PushDividendPayment.ts [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+      ).toEqual(true);
+      expect(addTransactionSpy.callCount).toEqual(1);
```
Comment (victorwiebe) [inline src/procedures/__tests__/PushDividendPayment.ts]: You are also missing one more line of code, which here would be
` expect(addTransactionSpy.getCall(0).lastArg.tag).toEqual(
        PolyTransactionTag.PushDividendPayment
      );`

See the function:
``` await this.addTransaction(dividendsModule!.pushDividendPaymentToAddresses, {
        tag: PolyTransactionTag.PushDividendPayment,
        // Only add resolver to the last transaction
        resolver:...
```
To see what we are getting at with this

File: src/procedures/__tests__/PushDividendPayment.ts [hunk 12]
```diff
@@ -0,0 +1,312 @@
...
+      const addTransactionSpy = spy(target, 'addTransaction');
+
+      // Real call
+      await target.prepareTransactions();
+
+      // Verifications
+      expect(
+        addTransactionSpy
+          .getCall(0)
+          .calledWith(erc20DividendsMock.getMockInstance().pushDividendPaymentToAddresses)
```
Comment (victorwiebe) [inline src/procedures/__tests__/PushDividendPayment.ts]: Same as above


> This is now only partially correct. You are missing 
> `erc20DividendsMock.mock('pushDividendPaymentToAddresses', Promise.resolve('PushDividendPaymentToAddresses'));` -> added somewhere before target.prepareTransactions
> 
> Where the resolve has a unique string. This is as per what I mentioned in slack today, let me know if you have more questions.
> 
> To test this, as the file is right now, you could change erc20DividendsMock on this line to ethDividendsMock, and it would still pass. That is because without the mock. the functions are all `undefined` type. If you add the code I mention above, you will see that you are no longer able to pass anything else into the calledWith, *only* this.

File: src/procedures/__tests__/PushDividendPayment.ts [hunk 13]
```diff
@@ -0,0 +1,312 @@
...
+      ).toEqual(true);
+      expect(addTransactionSpy.callCount).toEqual(2);
```
Comment (victorwiebe) [inline src/procedures/__tests__/PushDividendPayment.ts]: Same as above

> You are also missing one more line of code, which here would be
> expect(addTransactionSpy.getCall(0).lastArg.tag).toEqual( PolyTransactionTag.PushDividendPayment );
> 
> See the function:
> 
>         tag: PolyTransactionTag.PushDividendPayment,
>         // Only add resolver to the last transaction
>         resolver:...
> To see what we are getting at with this

File: src/procedures/__tests__/PushDividendPayment.ts [hunk 14]
```diff
@@ -0,0 +1,312 @@
...
+      const addTransactionSpy = spy(target, 'addTransaction');
+
+      // Real call
+      await target.prepareTransactions();
+
+      // Verifications
+      expect(
+        addTransactionSpy
+          .getCall(0)
+          .calledWith(ethDividendsMock.getMockInstance().pushDividendPaymentToAddresses)
```
Comment (victorwiebe) [inline src/procedures/__tests__/PushDividendPayment.ts]: Same as above (but eth instead of erc20)

This is now only partially correct. You are missing
ethDividendsMock.mock('pushDividendPaymentToAddresses', Promise.resolve('PushDividendPaymentToAddresses')); -> added somewhere before target.prepareTransactions

Where the resolve has a unique string. This is as per what I mentioned in slack today, let me know if you have more questions.

To test this, as the file is right now, you could change ethDividendsMock on this line to erc20DividendsMock, and it would still pass. That is because without the mock. the functions are all undefined type. If you add the code I mention above, you will see that you are no longer able to pass anything else into the calledWith, only this.

File: src/procedures/__tests__/PushDividendPayment.ts [hunk 15]
```diff
@@ -0,0 +1,312 @@
...
+      ).toEqual(true);
+      expect(addTransactionSpy.callCount).toEqual(1);
```
Comment (victorwiebe) [inline src/procedures/__tests__/PushDividendPayment.ts]: Same as above

> You are also missing one more line of code, which here would be
> expect(addTransactionSpy.getCall(0).lastArg.tag).toEqual( PolyTransactionTag.PushDividendPayment );
> 
> See the function:
> 
>         tag: PolyTransactionTag.PushDividendPayment,
>         // Only add resolver to the last transaction
>         resolver:...
> To see what we are getting at with this

File: src/procedures/__tests__/PushDividendPayment.ts [hunk 16]
```diff
@@ -0,0 +1,312 @@
...
+        factoriesMockedSetup,
+        params.symbol,
+        DividendType.Erc20,
+        0
+      )();
+
+      expect(
+        refreshStub.getCall(0).calledWithExactly(
+          DividendDistribution.generateId({
+            securityTokenId: SecurityToken.generateId({
```
Comment (victorwiebe) [inline src/procedures/__tests__/PushDividendPayment.ts]: You can define the securityTokenId in the beforeEach and call generateId there too. Saves a little bit of code declaring it gloablly

File: src/procedures/__tests__/PushDividendPayment.ts [hunk 17]
```diff
@@ -0,0 +1,312 @@
...
+        factoriesMockedSetup,
+        params.symbol,
+        DividendType.Eth,
+        1
+      )();
+
+      expect(
+        refreshStub.getCall(0).calledWithExactly(
+          DividendDistribution.generateId({
+            securityTokenId: SecurityToken.generateId({
```
Comment (victorwiebe) [inline src/procedures/__tests__/PushDividendPayment.ts]: You can define the securityTokenId in the beforeEach and call generateId there too. Saves a little bit of code declaring it gloablly

File: src/procedures/__tests__/PushDividendPayment.ts [hunk 18]
```diff
@@ -0,0 +1,312 @@
...
+              symbol: params.symbol,
+            }),
+            dividendType: DividendType.Erc20,
+            index: 0,
+          })
+        )
+      ).toEqual(true);
+      expect(await resolverValue).toEqual(undefined);
```
Comment (victorwiebe) [inline src/procedures/__tests__/PushDividendPayment.ts]: No await needed here

File: src/procedures/__tests__/PushDividendPayment.ts [hunk 19]
```diff
@@ -0,0 +1,312 @@
...
+          DividendDistribution.generateId({
+            securityTokenId: SecurityToken.generateId({
+              symbol: params.symbol,
+            }),
+            dividendType: DividendType.Eth,
+            index: 1,
+          })
+        )
+      ).toEqual(true);
+      expect(await resolverValue).toEqual(undefined);
```
Comment (victorwiebe) [inline src/procedures/__tests__/PushDividendPayment.ts]: No await needed here
